src/marks/raster.js

```diff
@@ -0,0 +1,146 @@
+import {applyAttr, applyDirectStyles, applyIndirectStyles, applyTransform, impliedString} from "../style.js";
+import {initializer} from "../transforms/basic.js";
+
+const defaults = {
+  ariaLabel: "raster",
+  stroke: null
+};
+
+export class Raster extends Mark {
+  constructor(data, options = {}) {
+    let {
+      width,
+      height,
+      x,
+      y,
+      // If X and Y are not given, we assume that F is a dense array of samples
+      // covering the entire grid in row-major order. These defaults allow
+      // further shorthand where x and y represent grid column and row index.
+      x1 = x == null ? 0 : undefined,
+      y1 = y == null ? 0 : undefined,
+      x2 = x == null ? width : undefined,
+      y2 = y == null ? height : undefined,
+      imageRendering,
+      pixelRatio = 1,
+      fill
+    } = options;
+    super(
+      data,
+      {
+        x: {value: x, scale: "x", optional: true},
+        y: {value: y, scale: "y", optional: true},
+        x1: {value: [+x1], scale: "x", filter: null},
+        y1: {value: [+y1], scale: "y", filter: null},
+        x2: {value: [+x2], scale: "x", filter: null},
+        y2: {value: [+y2], scale: "y", filter: null}
+      },
+      data == null && typeof fill === "function" ? sampleFill(options) : options,
+      defaults
+    );
+    this.width = width === undefined ? undefined : Math.floor(width);
+    this.height = height === undefined ? undefined : Math.floor(height);
+    this.pixelRatio = +pixelRatio;
+    this.imageRendering = impliedString(imageRendering, "auto");
+  }
+  render(index, scales, channels, dimensions, context) {
+    const {x: X, y: Y, fill: F} = channels;
+    if (!F) return;
+    const {x1: [x1], y1: [y1], x2: [x2], y2: [y2]} = channels; // prettier-ignore
+    const {document} = context;
+    const imageWidth = Math.abs(x2 - x1);
+    const imageHeight = Math.abs(y2 - y1);
+    const {
+      pixelRatio,
+      width = Math.round(imageWidth / pixelRatio),
+      height = Math.round(imageHeight / pixelRatio),
+      imageRendering
+    } = this;
+    const canvas = document.createElement("canvas");
+    canvas.width = width;
+    canvas.height = height;
+    const context2d = canvas.getContext("2d");
+    const image = context2d.createImageData(width, height);
+    const imageData = image.data;
+    if (X && Y) {
+      // If X and Y are given, then assign each sample to the corresponding
+      // pixel location. In the future, it would be better to allow different
+      // interpolation methods here, as the current approach will often lead to
+      // a sparse image when not every pixel has a corresponding sample.
+      const kx = width / imageWidth;
+      const ky = height / imageHeight;
+      for (const i of index) {
+        const xi = Math.floor((X[i] - x1) * kx);
+        if (xi < 0 || xi >= width) continue;
+        const yi = Math.floor((Y[i] - y2) * ky);
+        if (yi < 0 || yi >= height) continue;
+        const {r, g, b} = rgb(F[i]);
```

Comment:
I think it would be better to do the optimization earlier when the scale is applied, like how Yuri does here:

```js
// Convenience function to create a cached color interpolator that
// returns cached rgb objects, avoiding color string parsing.
cacheInterpolator = (interpolator, n = 250) =>
  d3.scaleQuantize(d3.quantize(pc => d3.rgb(interpolator(pc)), n))
```

https://observablehq.com/@twitter/density-plot@4159

But in any case I think we should do optimizations as follow-up, as you say.